### PR TITLE
support chat_model native on Intel Gaudi 

### PR DESCRIPTION
**Pull request by jeanyu-habana**

### CODE REVIEW

**Comment by ftian1** [comment]
This interface is used to provide langchain like api to access the remote OPEA services. I don't think it's a good idea to provide device specific interface. user should be invisible to such low level details.

**Comment by avinashkarani** [comment]
@jeanyu-habana, can we abstract Gaudi interface under OPEA abstraction and use HPU as argument. This way developers don't need to change their application to support Gaudi.

**Comment by jeanyu-habana** [comment]
> @jeanyu-habana, can we abstract Gaudi interface under OPEA abstraction and use HPU as argument. This way developers don't need to change their application to support Gaudi.

Current implementation of chat_model in OPEA is based on BaseChatOpenAI class, which means it forces the model to follow the OpenAI API format. The chat_model we are adding is to enable native support without depending on any API/Endpoint. Hiding the chat_model for native support under the non-native ChatOpea is confusing to the users. In fact, the users need to understand the differences and choose between two. 

Although APIs/Endpoints can be useful, there are use cases where native support is more efficient. For example, to take advantage of langchain's  GraphCypherQAChain and create a chatbot. In this case, the chat_model is a part of the chain, we need to have the chat_model run natively without having to create a separate API. We don't need another API just for each part of the chain. The chatbot (overall chain) is enabled as an API. 

Existing OPEA llm text_generation provides both native and non-native support. Below are explanations on why native support is better when performance optimization matters:
"LLM Native microservice uses [optimum-habana](https://github.com/huggingface/optimum-habana) for model initialization and warm-up, focusing solely on large language models (LLMs). It operates without frameworks like TGI/VLLM, using PyTorch directly for inference, and supports only non-stream formats. This streamlined approach optimizes performance on Habana hardware."
More details can be found here:
https://github.com/opea-project/GenAIComps/blob/main/comps/llms/src/text-generation/README_native.md

To summarize, I would argue that a native chat_model for Gaudi is necessary for use cases when chat_model is a part of the chain and performance is important. Adding native support does not stop users from using the existing OPEA chat_model if they don't mind the overhead of the OpenAI mandate.

**Comment by jeanyu-habana** [comment]
> This interface is used to provide langchain like api to access the remote OPEA services. I don't think it's a good idea to provide device specific interface. user should be invisible to such low level details.

Thanks for meeting with me. Just to summarize on our discussion, we agreed on the value of adding Gaudi langchain chat_models in OPEA. We are doing research to figure out the best place to add the support:  location/component/repo/etc. 

On 02/27/2025, after multiple conversations, Fang (@ftian1 ) and I agreed on renaming ChatGauti to ChatNative to avoid being specific to a device (hpu). However, within ChatNative we only support hpu for now.

**Comment by jeanyu-habana** [comment]
@ftian1 @avinashkarani Could you please review the changes per our discussions? Please let me know if there are additional comments before approval. Thank you.

**Review by avinashkarani — APPROVED** [review]

File: langchain_opea/native_chat_models.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+            args.use_hpu_graphs = True
+            args.use_kv_cache = True
+
+            pipe = GaudiTextGenerationPipeline(
+                args,
+                logger, 
+                use_with_langchain=True
+            )
+            hfpipe = HuggingFacePipeline(pipeline=pipe)
+            model, _, tokenizer, _= initialize_model(args, logger)
```
Comment (avinashkarani) [inline langchain_opea/native_chat_models.py]: Can we encapsulate the Gaudi and HF pipeline creation within ChatNative? This would provide a cleaner implementation, allowing developers to simply instantiate ChatNative with HPU as the device (along with additional arguments), which would internally handle pipeline creation.

Comment (jeanyu-habana) [reply langchain_opea/native_chat_models.py]: > ChatNative

Just to clarify, the code you referenced above is in the example of how to use ChatNative. The ChatNative itself is not directly tied to the hpu device. I updated the example to show hpu specific usage per your request:


            if args.device == "hpu":
                pipe = GaudiTextGenerationPipeline(
                    args,
                    logger,
                    use_with_langchain=True
                )
                hfpipe = HuggingFacePipeline(pipeline=pipe)
                model, _, tokenizer, _= initialize_model(args, logger)
            else:
                raise NotImplementedError(f"Only support hpu device now, device {args.device} not supported.")

            chat_model = ChatNative(
                temperature=0.1,
                llm=hfpipe,
                tokenizer=tokenizer
            )

Since ChatNative itself is not hpu specific, no update is implemented. ChatNative will work with any device/provider as long as the the llm/tokenizer (required by ChatNative) are defined/initialized for that device/provider.

Comment (ftian1) [reply langchain_opea/native_chat_models.py]: I agree with @avinashkarani 's comment. it's better to move/hide those Gaudi/HF pipeline func calls into ChatNative. with this way, user can get a clean API to use.

Comment (jeanyu-habana) [reply langchain_opea/native_chat_models.py]: > I agree with @avinashkarani 's comment. it's better to move/hide those Gaudi/HF pipeline func calls into ChatNative. with this way, user can get a clean API to use.

Okay, updated based on your comments. Please review. Thanks! @ftian1

File: langchain_opea/integrations/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+import torch
+from transformers import TextGenerationPipeline
+
+
+SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
+sys.path.append(os.path.dirname(SCRIPT_DIR))
+
+
+class GaudiTextGenerationPipeline(TextGenerationPipeline):
```
Comment (ftian1) [inline langchain_opea/integrations/pipeline.py]: suggest renaming it to _GaudiTextGenerationPipeline as it's just an internal class for use.

**Review by ftian1 — APPROVED** [review]
it's ok for me, just minor comment to suggest renaming GaudiTextGenerationPipeline to _GaudiTextGenerationPipeline

**Comment by jeanyu-habana** [comment]
> it's ok for me, just minor comment to suggest renaming GaudiTextGenerationPipeline to _GaudiTextGenerationPipeline

Based on our discussion, a better way is to use the GaudiTextGenerationPipeline defined in optimum-habana. So we can remove the local copy here.

**Review by ftian1 — APPROVED** [review]
